structuralcodes/materials/constitutive_laws.py

```diff
@@ -0,0 +1,204 @@
+
+    def get_stress(self, eps: ArrayLike) -> float:
+        """Return stress given strain."""
+        eps = np.asarray(eps)
+        return self._E * eps
+
+    def get_tangent(self, eps: ArrayLike) -> float:
+        """Return the tangent."""
+        return self._E
+
+
+class ElasticPlastic(ConstitutiveLaw):
+    """Class for elastic-plastic Constitutive Law."""
+
+    __materials__: t.Tuple[str] = (
+        'steel',
+        'rebars',
+    )
+
+    def __init__(
+        self,
+        E: float,
+        fy: float,
+        Eh: float = 0.0,
+        eps_su: t.Optional[float] = None,
+        name: t.Optional[str] = None,
+    ) -> None:
+        """Initialize an Elastic-Plastic Material."""
+        name = name if name is not None else 'ElasticPlasticLaw'
+        super().__init__(name=name)
+        if E > 0:
+            self._E = E
+        else:
+            raise ValueError('Elastic modulus E must be greater than zero')
+        self._fy = fy
+        self._Eh = Eh
+        self._eps_su = eps_su
+        self._eps_sy = fy / E
+
+    def get_stress(self, eps: ArrayLike) -> ArrayLike:
+        """Return the stress given strain."""
+        eps = np.atleast_1d(np.asarray(eps))
+        sig = self._E * eps
+        delta_sig = self._fy * (1 - self._Eh / self._E)
+        sig[sig < -self._fy] = eps[sig < -self._fy] * self._Eh - delta_sig
+        sig[sig > self._fy] = eps[sig > self._fy] * self._Eh + delta_sig
+        if self._eps_su is not None:
+            sig[eps > (self._eps_su * 1.01)] = 0
+            sig[eps < (-self._eps_su * 1.01)] = 0  # pylint: disable=E1130
+        return sig
+
+    def get_tangent(self, eps: ArrayLike) -> ArrayLike:
+        """Return the tangent for given strain."""
+        tol = 1.0e-6
+        if abs(eps) - self._eps_sy > tol:
+            return self._Eh
+        return self._E
+
+    def get_ultimate_strain(self) -> t.Tuple[float, float]:
+        """Return the ultimate strain (positive and negative)."""
+        return (self._eps_su, -self._eps_su)
+
+
+class ParabolaRectangle(ConstitutiveLaw):
+    """Class for parabola rectangle constitutive law."""
+
+    __materials__: t.Tuple[str] = ('concrete',)
+
+    def __init__(
+        self,
+        fc: float,
+        eps_0: float = -0.002,
+        eps_u: float = -0.0035,
+        n: float = 2.0,
+        name: t.Optional[str] = None,
+    ) -> None:
+        """Initialize an Elastic-Plastic Material."""
+        name = name if name is not None else 'ParabolaRectangleLaw'
+        super().__init__(name=name)
+        self._fc = -abs(fc)
+        self._eps_0 = -abs(eps_0)
+        self._eps_u = -abs(eps_u)
+        self._n = n
+
+    def get_stress(self, eps: ArrayLike) -> ArrayLike:
+        """Return the stress given strain."""
+        eps = np.atleast_1d(np.asarray(eps))
+        # Parabolic branch
+        sig = self._fc * (1 - (1 - (eps / self._eps_0)) ** self._n)
+        # Rectangle branch
+        sig[eps < self._eps_0] = self._fc
```

Comment:
Hi. Im a bit late to the party, but shouldn't the operator be flipped such that values where eps > self._eps_0 is set equal to self._fc?

Reply:
`eps_0` is negative, and forced negative in `__init__`, but we should also make sure that the user gets a heads up about this behaviour. Please add a docstring for the class :smiley: